### PR TITLE
implement pending nonce with UnconfirmedTx

### LINKED ISSUE #242: `EVMBackend.PendingTransactions` is not implemented
__Proposal:__ Implement it with `UnconfirmedTxs`.

__Current behavior:__ empty implementation

__Desired behavior:__ implement with `UnconfirmedTxs`

__Use case:__ To make `getTransactionCount(addr, 'pending')` works

### PR DESCRIPTION
**Pull request by yihuang**
Closes #242

Also simplified rpc query with `rpctypes.ContextWithHeight`.

## Description

It passes the python script test:

```
    w3 = web3.Web3(web3.providers.HTTPProvider("http://localhost:1317"))
    addr = w3.geth.personal.importRawKey(VALIDATOR_PRIV_KEY, "1")
    nonce1 = w3.eth.get_transaction_count(addr)
    tx = w3.eth.send_transaction({"from": addr, "to": COMMUNITY, "value": 1000})
    assert w3.eth.get_transaction_count(addr) == nonce1
    assert w3.eth.get_transaction_count(addr, "pending") == nonce1 + 1, 'invalid pending nonce'
    receipt = waittx(w3, tx)
    assert w3.eth.get_transaction_count(addr) == nonce1 + 1, 'invalid nonce'
```

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/namespaces/eth/api.go [hunk 1]
```diff
@@ -34,6 +31,7 @@ import (
 
 	"github.com/tharsis/ethermint/crypto/hd"
 	"github.com/tharsis/ethermint/ethereum/rpc/backend"
+	"github.com/tharsis/ethermint/ethereum/rpc/types"
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:34]: fix lint

File: ethereum/rpc/backend/backend.go [hunk 2]
```diff
@@ -294,8 +294,32 @@ func (e *EVMBackend) GetTransactionLogs(txHash common.Hash) ([]*ethtypes.Log, er
 
 // PendingTransactions returns the transactions that are in the transaction pool
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:295]: this should maybe state that only `MsgEthereumTx` are considered

File: ethereum/rpc/namespaces/eth/api.go [hunk 3]
```diff
@@ -985,7 +983,8 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
 // and have a from address that is one of the accounts this node manages.
 func (e *PublicAPI) PendingTransactions() ([]*rpctypes.RPCTransaction, error) {
 	e.logger.Debugln("eth_getPendingTransactions")
-	return e.backend.PendingTransactions()
+
+	return []*types.RPCTransaction{}, nil
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:987]: why empty?

Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:987]: To implement `eth_getPendingTransactions`, we have to do more queries to fetch block information, maybe do it in another issue.

File: ethereum/rpc/namespaces/eth/api.go (hunk 1 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:34]: done

File: ethereum/rpc/backend/backend.go (hunk 2 above)
Comment (yihuang) [reply ethereum/rpc/backend/backend.go:296]: I changed it to return `sdk.Tx`, the rpc side still filters `MsgEthereumTx` out, seems better this way.

File: ethereum/rpc/namespaces/eth/api.go (hunk 3 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go]: https://github.com/tharsis/ethermint/issues/244

File: ethereum/rpc/backend/backend.go [hunk 4]
```diff
@@ -294,8 +295,23 @@ func (e *EVMBackend) GetTransactionLogs(txHash common.Hash) ([]*ethtypes.Log, er
...
+		return nil, err
+	}
+
+	result := make([]*sdk.Tx, 0, len(res.Txs))
+	for _, txBz := range res.Txs {
+		tx, err := e.clientCtx.TxConfig.TxDecoder()(txBz)
+		if err != nil {
+			result = append(result, nil)
+			continue
+		}
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:310]: I really think we should log and return the error here, no transaction should fall under this case. If it does then there is something wrong

Comment (yihuang) [reply ethereum/rpc/backend/backend.go]: agree, done

File: ethereum/rpc/backend/backend.go [hunk 5]
```diff
@@ -294,8 +295,22 @@ func (e *EVMBackend) GetTransactionLogs(txHash common.Hash) ([]*ethtypes.Log, er
...
+	res, err := e.clientCtx.Client.UnconfirmedTxs(e.ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]*sdk.Tx, 0, len(res.Txs))
+	for _, txBz := range res.Txs {
+		tx, err := e.clientCtx.TxConfig.TxDecoder()(txBz)
+		if err != nil {
+			return nil, err
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:308]:
```suggestion
			e.logger.WithError(err).Debugln("tx decoding failed")
			return nil, err
```

**Review by fedekunze — APPROVED** [review]

**Pull request merged**